File: app/src/main/scala/farjs/app/filelist/zip/ZipEntry.scala
package farjs.app.filelist.zip

import farjs.app.util.DateTimeUtil

case class ZipEntry(
  parent: String,
  name: String,
  isDir: Boolean = false,
  size: Double = 0.0,
  datetimeMs: Double = 0.0,
  permissions: String = ""
)

object ZipEntry {

  private val itemRegex = """([d|-].+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+)""".r

  def fromUnzipCommand(output: String): List[ZipEntry] = {
    output.trim.split('\n').flatMap { line =>
      for {
        itemRegex(permissions, _, _, length, _, _, datetime, pathName) <- itemRegex.findFirstMatchIn(line)
      } yield {
        val path = pathName.stripSuffix("/")
        val lastSlash = path.lastIndexOf('/')
        val (parent, name) =
          if (lastSlash != -1) path.splitAt(lastSlash)
          else ("", path)

        ZipEntry(
          parent = parent,
          name = name.stripPrefix("/"),
          isDir = pathName.endsWith("/"),
          size = toDouble(length),
          datetimeMs = DateTimeUtil.parseDateTime(datetime),
          permissions = permissions
        )
      }
    }.toList
  }

  private def toDouble(s: String): Double = if (s.isEmpty) 0.0 else s.toDouble
}

File: archiver/zip/ZipEntry.mjs
/**
 * @import { FileListItem } from "@farjs/filelist/api/FileListItem.mjs"
 */
import { lazyFn, stripPrefix, stripSuffix } from "@farjs/filelist/utils.mjs";
import DateTimeUtil from "../DateTimeUtil.mjs";

/**
 * @typedef {FileListItem & {
 *  readonly parent: string;
 * }} ZipEntry
 */

/**
 * @param {string} parent
 * @param {string} name
 * @param {boolean} [isDir]
 * @param {number} [size]
 * @param {number} [datetimeMs]
 * @param {string} [permissions]
 * @returns {ZipEntry}
 */
function ZipEntry(parent, name, isDir, size, datetimeMs, permissions) {
  const entryTimeMs = datetimeMs ?? 0;
  return {
    parent,
    name,
    isDir: isDir ?? false,
    isSymLink: false,
    size: size ?? 0,
    atimeMs: entryTimeMs,
    mtimeMs: entryTimeMs,
    ctimeMs: entryTimeMs,
    birthtimeMs: entryTimeMs,
    permissions: permissions ?? "",
  };
}

/**
 * @param {string} output
 * @returns {ZipEntry[]}
 */
ZipEntry.fromUnzipCommand = (output) => {
  return output
    .trim()
    .split("\n")
    .map((line) => {
      const regexRes = itemRegex().exec(line);
      if (regexRes) {
        const permissions = regexRes[1];
        const size = parseFloat(regexRes[4]);
        const datetime = regexRes[7];
        const pathName = regexRes[8];

        const path = stripSuffix(pathName, "/");
        const lastSlash = path.lastIndexOf("/");
        const [parent, name] =
          lastSlash !== -1
            ? [path.substring(0, lastSlash), path.substring(lastSlash + 1)]
            : ["", path];

        return ZipEntry(
          parent,
          stripPrefix(name, "/"),
          pathName.endsWith("/"),
          size,
          DateTimeUtil.parseDateTime(datetime),
          permissions,
        );
      }
      return undefined;
    })
    .filter((_) => _ !== undefined);
};

/** @type {() => RegExp} */
const itemRegex = lazyFn(
  () => /([d|-].+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+?)\s+(.+)/,
);

export default ZipEntry;
